data/pools/transactionPool.go

```diff
@@ -179,6 +179,44 @@ func (pool *TransactionPool) Remember(t transactions.SignedTxn) error {
 	defer pool.mu.Unlock()
 
 	err := pool.test(t)
+	// Check for a race condition when a new block was added
+	// to our ledger, but hasn't propagated to OnNewBlock()
+	// yet.
+	switch errt := err.(type) {
+	case transactions.TxnDeadError:
+		if errt.Round >= errt.FirstValid {
+			// This is not an error due to the round being too old.
+			break
+		}
+
+		if pool.pendingBlockEvaluator == nil {
+			// We don't have a block evaluator at all; something
+			// else went wrong, and this isn't the place to fix
+			// it.  (The next OnNewBlock callback might recover.)
+			break
+		}
+
+		latest := pool.ledger.Latest()
+		if pool.pendingBlockEvaluator.Round() > latest {
```

Comment:
this is all evaluated within the same lock ( pool.mu ). Is it really possible ? i.e. it should have been evaluated in test().

Reply:
You're suggesting to move this new blob of code (retrying, etc) into `test()`?  That's probably a good idea; thanks!